main/src/ca/uwaterloo/flix/language/phase/Safety.scala

```diff
@@ -351,20 +350,105 @@ object Safety {
+      * @param clazz the Java class.
+      */
+    case class NonJavaType(tpe: Type, clazz: java.lang.Class[_]) extends JavaSubtypeResult
+
+    /**
+      * @param tpe the type variable.
+      */
+    case class TypeVariable(tpe: Type) extends JavaSubtypeResult
+  }
+
+  /**
+    * Returns true if `tpe1` and `tpe2` are both Java types
+    * and `tpe1` is a subtype of `tpe2`.
+    * Note that `tpe1` is also allowed to be a Flix string
+    * or BigInt/BigDecimal while `tpe2` is a supertype of this.
+    */
+  private def isJavaSubtypeOf(tpe1: Type, tpe2: Type)(implicit flix: Flix): JavaSubtypeResult = (tpe1.baseType, tpe2.baseType) match {
+
+    case (Type.Cst(TypeConstructor.Native(left), _), Type.Cst(TypeConstructor.Native(right), _)) =>
+      if (right.isAssignableFrom(left)) JavaSubtypeResult.Castable else JavaSubtypeResult.NonCastable
+
+    case (Type.Cst(TypeConstructor.Str, _), Type.Cst(TypeConstructor.Native(right), _)) =>
+      if (right.isAssignableFrom(classOf[java.lang.String])) JavaSubtypeResult.Castable else JavaSubtypeResult.NonCastable
+
+    case (Type.Cst(TypeConstructor.BigInt, _), Type.Cst(TypeConstructor.Native(right), _)) =>
+      if (right.isAssignableFrom(classOf[java.math.BigInteger])) JavaSubtypeResult.Castable else JavaSubtypeResult.NonCastable
+
+    case (Type.Cst(TypeConstructor.BigDecimal, _), Type.Cst(TypeConstructor.Native(right), _)) =>
+      if (right.isAssignableFrom(classOf[java.math.BigDecimal])) JavaSubtypeResult.Castable else JavaSubtypeResult.NonCastable
+
+    case (Type.Var(_, _), _) =>
+      JavaSubtypeResult.TypeVariable(tpe1)
+
+    case (_, Type.Var(_, _)) =>
+      JavaSubtypeResult.TypeVariable(tpe2)
+
+    case (Type.Cst(TypeConstructor.Native(clazz), _), _) =>
+      JavaSubtypeResult.NonJavaType(tpe2, clazz)
+
+    case (_, Type.Cst(TypeConstructor.Native(clazz), _)) =>
+      JavaSubtypeResult.NonJavaType(tpe1, clazz)
+
+    case _ => JavaSubtypeResult.NonCastable
+  }
+
+  /**
+    * Returns a list of errors if the the upcast is invalid.
+    */
+  private def checkUpcastSafety(exp: Expression, tpe: Type, renv: RigidityEnv, loc: SourceLocation)(implicit flix: Flix): List[SafetyError] = {
+    val tpe1 = Type.eraseAliases(exp.tpe)
+    val tpe2 = Type.eraseAliases(tpe)
+    if (isSubtypeOf(tpe1, tpe2, renv))
+      Nil
+    else
+      UnsafeUpcast(exp.tpe, tpe, loc) :: Nil
+  }
+
+  /**
+    * Returns a list of errors if the the supercast is invalid.
+    */
+  private def checkSupercastSafety(exp: Expression, tpe: Type, loc: SourceLocation)(implicit flix: Flix): List[SafetyError] = {
+    val tpe1 = Type.eraseAliases(exp.tpe)
+    val tpe2 = Type.eraseAliases(tpe)
+    isJavaSubtypeOf(tpe1, tpe2) match {
+      case JavaSubtypeResult.Castable => Nil
+      case JavaSubtypeResult.NonCastable => UnsafeSupercast(exp.tpe, tpe, loc) :: Nil
+      case JavaSubtypeResult.NonJavaType(t, c) if t == tpe1 => FromNonJavaTypeSupercast(exp.tpe, c, loc) :: Nil
```

Comment:
Can't the returned value just tell which side is non-java instead of checking like this?